### PR TITLE
Pass environment variables to LSP

### LINKED ISSUE #214: Set project specific environment variables for lsp server
To build my project, i need to set some environment-variables (which can't be global since they are project-specific). Is this possible so that the HIE worker can build the project? Some VS-Code extensions allow the setting of specific environment-variables.

### PR DESCRIPTION
**Pull request by jacobprudhomme**
## Summary
Made it so that the user can define environment variables that will then be passed to the LSP upon execution. Resolves #214 

## Changes
* Added extension settings to define environment variables
* Passed environment variables to LSP

## Notes
I had a bit of trouble testing this, I'm not sure if the environment variables actually get passed. Is there a test scenario someone can help me come up with to test this?

### CODE REVIEW

**Comment by jneira** [comment]
Hi many thanks for the pr, loosk great overall (will commentabout some details)

> I had a bit of trouble testing this, I'm not sure if the environment variables actually get passed. Is there a test scenario someone can help me come up with to test this?

Well, to test this manually we could change a environment variable hls specific, like `HIE_BIOS_CACHE`, and check hls start to write to that path instead to `~/.cache/hie-bios`.

It could be done in automatic tests, reproducing those steps and wait a reasonable amount of time to check the directory is created.

File: package.json [hunk 1]
```diff
@@ -144,6 +144,12 @@
           "default": "",
           "markdownDescription": "Pass additional arguments to the language server."
         },
+        "haskell.envVars": {
```
Comment (jneira) [inline package.json:147]: what about 
```suggestion
        "haskell.serverEnvironment": {
```
?

in the cmake extension it is named `environment` but prefixing it with server we make clear we are using for it and the option would be showed near to `haskell.serverExtraArgs`

File: package.json [hunk 2]
```diff
@@ -144,6 +144,12 @@
...
+          "scope": "resource",
+          "type": "object",
+          "default": {},
+          "markdownDescription": "Define environment variables for compilation/execution."
```
Comment (jneira) [inline package.json:151]: We are passing the env vars for the lsp server (haskell-language-server) at runtime so not sure if `for compilation/execution` would be accurate 
What about 
```suggestion
          "markdownDescription": "Define environment variables for the language server."
```

**Comment by jneira** [comment]
Another improvement could be add a log output showing the environment vars added.

**Comment by jacobprudhomme** [comment]
> Well, to test this manually we could change a environment variable hls specific, like `HIE_BIOS_CACHE`, and check hls start to write to that path instead to `~/.cache/hie-bios`.

Good news! It looks like everything works when testing manually!

> It could be done in automatic tests, reproducing those steps and wait a reasonable amount of time to check the directory is created.

Unfortunately, while trying to write an automatic test for this, the directory doesn't show up, even after setting the timeout to a minute. I can't yet debug the tests to find out why, because I get `Error: Cannot find module '(...)/vscode-haskell/out/test'` when I try to launch debugging for the Extension Tests profile. Is this something you've encountered before?

> Another improvement could be add a log output showing the environment vars added. You already have the logger at hand to acomplish it.

For this, do you want me to log simply that the environment variables have been added, or do you want me to log the list of all environment variables added?

**Comment by jneira** [comment]
> Good news @jneira! It looks like everything works when testing manually!

Great!

> Unfortunately, while trying to write an automatic test for this, the directory doesn't show up, even after setting the timeout to a minute. I can't yet debug the tests to find out why, because I get Error: Cannot find module '(...)/vscode-haskell/out/test' when I try to launch debugging for the Extension Tests profile. Is this something you've encountered before?

debug normal code works for me but i did not try test code, will take a look

> For this, do you want me to log simply that the environment variables have been added, or do you want me to log the list of all environment variables added?

I think show the actual env vars would be useful, to double check the final value, as it could be set at global scope or per project and to be included in the logs users send us when there are problems.

I just opened a pr that logs all the env vars in case could be useful for doing that: https://github.com/haskell/vscode-haskell/pull/495/files

**Comment by jacobprudhomme** [comment]
@jneira where can I view these logs? I have looked around at the extension host logs and the logfile that the extension outputs, but cannot find any occurrence of the logger call I added, nor of the ones surrounding it (for example, the ones on [lines 236-237](https://github.com/haskell/vscode-haskell/blob/3f167498d1abcb617faf00d3d0591c5ef0a1abda/src/extension.ts#L236-L237)). I have also set the log level to `debug`, but that doesn't seem to have changed anything.

**Comment by jneira** [comment]
The log is in a specific buffer in the output panel, called `haskell (dirName)`:

imagen

**Comment by jacobprudhomme** [comment]
Got it! For some reason it wasn't showing it before, but it is now.

**Comment by jacobprudhomme** [comment]
@jneira it seems that the cache directory only gets created for stack and cabal projects. I tried writing a basic `stack.yaml`/`<project>.cabal` file to the root of the test workspace, hoping that was all that was needed to generate the cache directory. Unfortunately, I've learned HLS will actually try to run stack/cabal, upon detecting the type of project it is. If the necessary build tool is not available, the subsequent tests fail. I don't think it's ideal to introduce the necessity of having stack or cabal just to run this one test, though.

Is there maybe an environment variable that would be simpler to test automatically? I couldn't find any documentation on what other configuration HLS accepts via environment variables.

**Comment by jneira** [comment]
> Is there maybe an environment variable that would be simpler to test automatically? I couldn't find any documentation on what other configuration HLS accepts via environment variables.

hmm ghcide honours the xdg standard for its cache so if we change `XDG_CACHE_HOME` to a known location i think it will write something in it even only using ghc

**Comment by jacobprudhomme** [comment]
@jneira sorry to bother you so much! You were right, setting `XDG_CACHE_DIR` to something else worked. Unfortunately, now I'm really stumped.

When I run the tests, even though I _see_ the directory being created when the test window opens up, [my test case](https://github.com/haskell/vscode-haskell/blob/9dfda3c8442a1d8e9292c2ab67b1690080cc3952/test/suite/extension.test.ts#L103-L109) still waits 30 seconds and then declares it has not found the directory. I have no clue why this would be happening.

**Comment by jneira** [comment]
> When I run the tests, even though I _see_ the directory being created when the test window opens up, [my test case](https://github.com/haskell/vscode-haskell/blob/9dfda3c8442a1d8e9292c2ab67b1690080cc3952/test/suite/extension.test.ts#L103-L109) still waits 30 seconds and then declares it has not found the directory. I have no clue why this would be happening.

Hi, dont worry, the pr is looking great. I think the directory is already created due to previous timeouts when the test case start, so the watcher is not notificating its creation. Maybe a simple existence test reading directly from the file system will work (with a comment noting cause we are not using the watcher).

**Comment by jacobprudhomme** [comment]
Alright, that seems to have settled everything!

**Comment by jneira** [comment]
The linux fail is not related with this pr and i hit it in #495. So i am gonna merge as is, many many thans for work on this

**Review by jneira — APPROVED** [review]

**Comment by jacobprudhomme** [comment]
Thank you for all the help in making my first changes to the project!

**Pull request merged**